### PR TITLE
Add Dockerfile to run Obal as a container

### PR DESCRIPTION
**Pull request by ehelms**
WIll require https://github.com/theforeman/obal/pull/159

Note I have not fully built out the Travis automatic deployment aspect as I want to ensure the build direction (i.e. from PyPi assets) is the direction we want to go. I wasn't sure how to handle credentials within the script deployer yet as well.

### CODE REVIEW

**Comment by ehelms** [comment]
@evgeni From your experience, do you know if deploy stages are sequential?

**Comment by evgeni** [comment]
@ehelms on Travis? I don't *know*, but I *think* they are, as the parallelism in Travis comes from the branches (in our case Python versions) and the rest is executed more or less "as a script"

File: .travis.yml [hunk 1]
```diff
@@ -14,17 +14,24 @@ install:
...
-    python: '3.6'
+  - provider: pypi
+    user: theforeman
+    password:
+      secure: HmlFYR1Nn325ZYPTWOorcw2qSQNrpxOiEmi3sd3EhVNlk0sR0582ujNY2cLquLxQkFFs3cjDdUVz1FNHC0PX5nOSbA7P0Ks4xZoU2MeXq+HLlreB2w2qcp2XfJXbSFh8ZfqoMCw83lsHkTBwkCYpy/jF6FNUSiP0hWp0aYNMVKwaLB2V8YggkyYStUqtLyq8BlKP1bxWqZTjVia9sk5ZpzG+sDMC/rzwlYDRzk6u9UFITmav6ONCFIsiwBuOTa9KcON21PQGJOBnP3HJB3iyPcGKI/3Ln5aUKVn36tN5hNu3s8VkqVumpxZir5qpqplsJdpY9tRwUTGyVP+CVA8wjxJYP3ziRQpJeHXVdDQ3pU1t9uOPp5Zgfp1Xbbl5QhcAQkHedQ27Xc3WlKdgXZO4dQpTnuFPW71j/J4EM29M1/0RecSYrVvefqWk7ZhIgX4VHydzM9n7VJcIGjXkY5JIWTgEwbXsw9E58iktefYy7+q/z059e0m3+vLJmuWIE+eucmN9pWdunS6GEL8fq9XDL5egtf0c//+mrjHj4u90+8w+TjzdVduqQyQVI3MZ68eM+Wlz0ChjOk32l2N9CmNjGXRFog2YbazZaJT0Z+mRMUlyQUNr6nzC4smsqX8UqzrdHI/g6s9klNoNWdjTF4afewElkQtX9NT0m5SVWAr12hg=
+    on:
+      tags: true
+      python: '3.6'
+  - provider: script
+    script: bash .travis/deploy-container.sh
```
Comment (evgeni) [inline .travis.yml]: the file below is only called `.travis/deploy.sh`

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,20 @@
+FROM centos:7
```
Comment (evgeni) [inline Dockerfile]: given this is not strictly useful *inside* the python package, I think this file should be excluded from the MANIFEST (contrary to what `check-manifest` suggests)

File: README.md [hunk 3]
```diff
@@ -10,3 +10,16 @@ All `obal` actions should also work with plain Ansible when called like `ansible
...
+
+Obal and all it's required packages are available in a container that can be used locally or in build environments. The users Koji credentials and configuration must be mounted into the container alongside of mounting the packaging project into `/opt/packaging` to work. Note the examples below assume SELinux is disabled.
+
+With `podman`:
+
+    podman run -v `pwd`:/opt/packaging -v ~/.koji:/root/.koji obal:latest scratch katello
+
+With `docker`:
+
+		docker run -v `pwd`:/opt/packaging -v ~/.koji:/root/.koji obal:latest scratch katello
```
Comment (evgeni) [inline README.md]: The indentation is off, but I wonder if we really need two identical examples with a different binary name? I mean, `podman` is designed to be a drop-in replacement for `docker`, right?

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+		yum -y install epel-release && \
+		yum -y install python2-pip && \
+    yum clean all
+
+ENV LANG=en_US.UTF-8
+ENV LANGUAGE=en_US.UTF-8
+ENV LC_ALL=en_US.UTF-8
+ENV PYTHONUNBUFFERED=0
+
+RUN pip install obal
```
Comment (ekohl) [inline Dockerfile]: Does this mean it installs ansible from pip?

Comment (ehelms) [reply Dockerfile]: Yes.

**Comment by ehelms** [comment]
Updated a bit

File: README.md [hunk 5]
```diff
@@ -10,3 +10,11 @@ All `obal` actions should also work with plain Ansible when called like `ansible
...
 - `python` (2 or 3)
 - `ansible`
+
+## Using Obal via Container
+
+Obal and all it's required packages are available in a container that can be used locally or in build environments. The users Koji credentials and configuration must be mounted into the container alongside of mounting the packaging project into `/opt/packaging` to work. Note the examples below assume SELinux is disabled.
+
+To run:
+
+    podman run -v `pwd`:/opt/packaging -v ~/.koji:/root/.koji obal:latest scratch katello
```
Comment (ekohl) [inline README.md]: It would be good to mention that podman is a drop in for docker. I didn't know it was a drop in.

**Comment by ehelms** [comment]
The `MANIFEST` change was causing tox to fail at the start.

File: .travis.yml [hunk 6]
```diff
@@ -14,17 +14,24 @@ install:
...
+    on:
+      python: '3.6'
 script:
   - tox
 addons:
   apt:
     packages:
     - git-annex
     - rpm
+end:
```
Comment (ekohl) [inline .travis.yml]: env?

File: .travis.yml [hunk 7]
```diff
@@ -14,17 +14,24 @@ install:
...
+  global:
+    - secure: PUT_ENCRYPTED_HERE
```
Comment (ekohl) [inline .travis.yml]: Is this needed? The docs suggest that if you set it via the CLI then it's already available on every build.

Comment (ehelms) [reply .travis.yml]: This reads like this step will add it to the .travis.yml file: https://docs.travis-ci.com/user/environment-variables#encrypting-environment-variables

File: MANIFEST.in [hunk 8]
```diff
@@ -3,3 +3,5 @@ include LICENSE
 include .coveragerc
 include .pylintrc
 recursive-include rules *.py
+exclude Dockerfile
+exclude .travis
```
Comment (evgeni) [inline MANIFEST.in]: I also had to add `recursive-exclude .travis *` for that to work, as `exclude` is not recursive

File: Dockerfile [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+RUN obal setup
+
+RUN mkdir /opt/packaging
```
Comment (evgeni) [inline Dockerfile]: should this be `mkdir -p` to avoid problems if `/opt` shall ever vanish from the base?

File: Dockerfile [hunk 10]
```diff
@@ -0,0 +1,20 @@
+FROM centos:7
+
+RUN echo "tsflags=nodocs" >> /etc/yum.conf && \
```
Comment (evgeni) [inline Dockerfile]: indentation is off

**Comment by ehelms** [comment]
Was there an easy way to test that it will actually deploy properly in the Travis environment? Or lots of trial and error?

**Comment by evgeni** [comment]
> Was there an easy way to test that it will actually deploy properly in the Travis environment? Or lots of trial and error?

When I added the pypi deploy, I did not find anything we could do for a dry run :/

**Comment by ekohl** [comment]
For containers you can also let it deploy on a branch so you don't spam master.

File: MANIFEST.in [hunk 11]
```diff
@@ -3,3 +3,5 @@ include LICENSE
 include .coveragerc
 include .pylintrc
 recursive-include rules *.py
+exclude Dockerfile
+recursive-exclude .travis
```
Comment (ehelms) [inline MANIFEST.in]: any idea to get this to work?

**Comment by ehelms** [comment]
Blocked by https://github.com/theforeman/obal/issues/163

File: Dockerfile [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+    yum -y install epel-release && \
+    yum -y install python2-pip && \
+    yum clean all
+
+ENV LANG=en_US.UTF-8
+ENV LANGUAGE=en_US.UTF-8
+ENV LC_ALL=en_US.UTF-8
+ENV PYTHONUNBUFFERED=0
+
+RUN pip install obal
```
Comment (ekohl) [inline Dockerfile]: There's an issue here that this depends on the previous action: deployment to pypi. Chances are you're always building an old version in the container.

**Comment by evgeni** [comment]
Hmm. Can we instead pip install from a git link that includes the tag?

**Comment by ehelms** [comment]
I have this mostly working except for the credentials part through secrets. That part seems to be unclear from all docs I have read.

File: .travis.yml [hunk 13]
```diff
@@ -14,17 +14,26 @@ install:
...
+      branch: 'add-dockerfile'
 script:
   - tox
 addons:
   apt:
     packages:
     - git-annex
     - rpm
+env:
+  global:
```
Comment (ekohl) [inline .travis.yml]: Is this global part needed? Can't you set the env on the actual deploy?
```yaml
  - provider: script
    script: bash .travis/deploy-container.sh 0.2.0
    on:
      python: '3.6'
      branch: 'add-dockerfile'
    env:
      - secure: ...
      - secure: ...
```

Comment (ehelms) [reply .travis.yml]: Have you seen an example of that? I couldn't find the docs.

**Comment by ehelms** [comment]
So I've reached a point where I am not clear if its the credentials themselves thats the issue or how the credentials are being supplied. What do y'all think we should do to proceed? 

When I have tested this it mostly seems like either the parameters are not being supplied or they are incorrect especially this line -- https://github.com/theforeman/obal/pull/160/files#diff-6d3d682d6f98362acd0e77f23b9d7de5R5

Here is an example -- https://travis-ci.org/ehelms/obal/jobs/469167949

**Comment by ehelms** [comment]
I've removed the 'push' part of this to a follow up PR, and added a test build to the scripts section. Please re-review!

File: .travis.yml [hunk 14]
```diff
@@ -13,16 +13,20 @@ install:
...
-script:
-  - tox
+  - provider: pypi
+    user: theforeman
+    password:
+      secure: HmlFYR1Nn325ZYPTWOorcw2qSQNrpxOiEmi3sd3EhVNlk0sR0582ujNY2cLquLxQkFFs3cjDdUVz1FNHC0PX5nOSbA7P0Ks4xZoU2MeXq+HLlreB2w2qcp2XfJXbSFh8ZfqoMCw83lsHkTBwkCYpy/jF6FNUSiP0hWp0aYNMVKwaLB2V8YggkyYStUqtLyq8BlKP1bxWqZTjVia9sk5ZpzG+sDMC/rzwlYDRzk6u9UFITmav6ONCFIsiwBuOTa9KcON21PQGJOBnP3HJB3iyPcGKI/3Ln5aUKVn36tN5hNu3s8VkqVumpxZir5qpqplsJdpY9tRwUTGyVP+CVA8wjxJYP3ziRQpJeHXVdDQ3pU1t9uOPp5Zgfp1Xbbl5QhcAQkHedQ27Xc3WlKdgXZO4dQpTnuFPW71j/J4EM29M1/0RecSYrVvefqWk7ZhIgX4VHydzM9n7VJcIGjXkY5JIWTgEwbXsw9E58iktefYy7+q/z059e0m3+vLJmuWIE+eucmN9pWdunS6GEL8fq9XDL5egtf0c//+mrjHj4u90+8w+TjzdVduqQyQVI3MZ68eM+Wlz0ChjOk32l2N9CmNjGXRFog2YbazZaJT0Z+mRMUlyQUNr6nzC4smsqX8UqzrdHI/g6s9klNoNWdjTF4afewElkQtX9NT0m5SVWAr12hg=
+    on:
+      tags: true
+      python: '3.6'
+  - provider: script
```
Comment (evgeni) [inline .travis.yml]: shouldn't this deploy step go away too, as travis cannot deploy w/o having the credentials?

**Comment by evgeni** [comment]
the build failure is unrelated

**Pull request merged**